Add support for Cron Job definitions

horiajurcut opened the pull request:
We're adding Cron Job support in Encore. You can define a cron job at the package level using the following syntax:

```
/*
  In this example, we are defining a cron job, together with
  a random string.
  
  Come join the Encore revolution https://encore.dev/
  
  var _, bd = cron.NewJob("cronfood", cron.JobConfig{
	  Name:     "Cron Food Bar",
	  Schedule: "* * * * 5",
	  Endpoint: Cron,
  }), "barfoo"
  
  We are being very explicit, and we are also giving a description
  that includes a code snippet.
*/
var _, _ = cron.NewJob("supercronjob", cron.JobConfig{
	Name:     "Super Cron Job",
	Schedule: "45 23 * * 5",
	Endpoint: Cron,
}), "barfoo"
```

marcuskohlberg committed:
Update description (#108)

eandre committed:
compiler/internal/codegen: update config handling

Migrate to new config format while we're at it.

horiajurcut committed:
compiler/internal/codegen: tweak config

horiajurcut committed:
compiler/internal/codegen: fix syntax error in codegen

eandre committed:
compiler/internal/codegen: add sql database parsing

eandre committed:
compiler/internal/codegen: simplify config

eandre committed:
cli/daemon/sqldb: increase shared mem cap (#109)

The local development database was provisioned with
a very conservative, default memory limit of 16MB.

Change this to 1GB to support larger databases and
more complex queries.

eandre committed:
compiler/internal/codegen: support remaining types (#110)

Add support for the few additional types that were overlooked in the
implementation of the type decoder for query strings.

ValeriaVG committed:
docs: how to share db between services (#111)

* docs: how to share db between services

* docs: add how-to/share-db.. to sidebar

horiajurcut committed:
cli/daemon: database improvements (#113)

Various improvements to database handling.

Fixes #49, #98, #107

eandre committed:
cli: cleanup remote client

eandre committed:
add ability to override config dir

eandre committed:
cli/internal/platform: fix auth

The new platform client mistakenly did not always propagate
auth tokens.

horiajurcut committed:
implement login with auth key

horiajurcut committed:
compiler: cli: fix tests to match new cfg

horiajurcut committed:
parser: parse cron job definitions

eandre committed:
compiler: rewrite config handling to match new format

eandre committed:
cli/cmd/encore: restart daemon on config changes

eandre committed:
cli/daemon: fix runtime configuration

horiajurcut committed:
remove pretty print

eandre committed:
cli/daemon: improve error message when port is in use

ValeriaVG committed:
docs: document path conflicts & how to avoid them (#115)

horiajurcut committed:
cli/daemon/dash/dashapp: bump up vite version

horiajurcut committed:
cron: expand cron job definition

horiajurcut committed:
parser: write a test

horiajurcut committed:
parser: change cron job definition

DomBlack committed:
compiler/runtime/runtime/config: align config with runtime

This commit aligns the `config.Runtime` struct to match that as found in the
runtime package.

DomBlack committed:
all: organize imports

DomBlack committed:
cli: support alternative binaries

DomBlack committed:
Update cobra version to add auto completions

eandre committed:
all: improve database setup

horiajurcut committed:
parser: new cron job definition

horiajurcut committed:
parser: change description to doc

horiajurcut committed:
parser: extend functionality

horiajurcut committed:
remove whitespaces

DomBlack committed:
[ENC-308] Get CLI working in `encoredev` builds

This comimt adds two more varaibles we can override during dev builds of `encore`
such that we can point it and local development copies of the Go runtime and Encore runtimes

DomBlack commented on parser/parser.go:
```diff
@@ -275,95 +285,11 @@ func (p *parser) parseReferences() {
 	for _, pkg := range p.pkgs {
```
Could we use a call to `ast.Walk` in the future rather than the many nested loops?

DomBlack committed:
[ENC-307] Add support for resolving generic names

This commit adds support to the names package so we can resolve types which use
more than 1 generic type (i.e. `X[A, B]`) as this uses a new ast type which we
previously didn't support.

To make this backwards compitable, I've refactored the names package and introduced
the concept of "language levels" which allow us to add logic per-language level, which
we can use the built in language level tags to enable/disable those source files.

Meaning this change will still compile against Go 1.16 even though it's using a type
only avaliable in the Go 1.18 version of AST package.

DomBlack committed:
[ENC-307] Generic Response Types (#3)

This commit gets generic response types working under Go 1.18 and
allows the system to return a response struct with a generic field
within it.

This commit also adds the ability for the encore parser to understand
and obey `//go:build` tags, which helps with things like ARCH, OS,
language level or tooling tags filters on files - preventing us trying
to parse a file we can't or shouldn't.

DomBlack committed:
[ENC-309] Add support for Generic Input Types (#4)

* [ENC-309] Add support for Generic Input Types

This commit updates the code generation path so we can support
generic types as direct parameter inputs to RPC endpoints, such
as:

```go
type Optional[T any] struct {
    Present bool
    Value   T
}

//encore:api public
func MyAPI(ctx context.Context, params: Optional[string]) (string, error) {
    if params.Present {
        return fmt.Sprintf("Hello %s", params.Value), nil
    } else {
        return "Hey random person!", nil
    }
}
```

Note: I've also commited IntelliJ/GoLand run configurations for easily starting
the daemon or run commands locally.

* Addressing PR comments

DomBlack committed:
[ENC-310] Update DashApp to Support Generics (#5)

* [ENC-310] Add Support for Generics Client Side

This commit converts `est.Param` to contain a `*schema.Type` rather
than a `*schema.Decl`, as we now need to know what type arguments have
been given to initialise a type, where as a `Decl` is only the raw
underlying delecation of that type.

I've then updated the Go based go generator for a typescript client
to have full support for generics. While I was there I refactored
how the test is run, which makes it easier to both edit the input/output
of the test, but also add more languages in the future by simply adding an
`expected_langname.ext` file to the `testdata` folder.

This commit starts to add support for generics on the client side by
updating the proto code generator to also generate typescript files,
which we can then include directly from the dashapp. This minimises
the risk of the two systems (Go + Typescript) becoming different in
terms of their representation of the Encore app.

* [ENC-310] Update the dashapp to support generics

This commit updates the `dashapp` so it works with generic types.
While I was here I removed some duplication of code in the JSON
default parameters for requests. As part of unifiying the dupliated
code I've moved the SchemaView over to using `CodeMirror` as well
so we get syntax highlighting in the API docs.

I've also updated the response renderers in both `RPCCaller` and
the trace overview to also render the payloads with syntax highlighting.

* Addressing PR comments

eandre committed:
cli/cmd/encore: support verbose logging

This adds a global "-v" flag that enables debug logging.
It can be specified twice to enable trace logging.

eandre committed:
cli/internal/platform: add trace logs

Log requests and responses to and from the platform when trace
logging is enabled (verbosity level 2 and above).

DomBlack committed:
[ENC-312] Add missing optional to `AuthHandler` (#6)

The meta data type did not flag the AuthHandler as optional.
This commit fixes that.

eandre committed:
compiler/runtime/runtime/config: fix secrets loading

Load secrets using the new config mechanism.

eandre committed:
compiler: use reflect for auth data checks

This is more robust against slightly different string representations
of type names.

eandre committed:
compiler/internal/codegen: use external test package

To work around potential import cycles due to our new use
of reflect.TypeOf for auth data, move the test config into
an external "foo_test" package.

eandre committed:
.github/workflows: run build with Go 1.17

DomBlack committed:
Disable generic tests when not Go 1.18 (#119)

This commit uses Go build tags to disable two tests which now
rely on Go 1.18 generics which are causing test failures when we
test on Go 1.17 or below.

eandre committed:
cli: implement log streaming using websockets (#121)

* cli: move cli/daemon/internal/appfile to cli/internal/appfile

In preparation for moving logs from being proxied through the
daemon into being handled directly by the client.

* cli: proto: implement log streaming in client

Move this to the client to simplify keepalive handling
and needless boilerplate.

eandre committed:
pgproxy: add SCRAM support, simplify implementation (#120)

* pgproxy: simplify implementation

Instead of vendoring parts of pgproto3 simply
call out to these libraries directly.

Also add support for SCRAM authentication, based on
github.com/jackc/pgconn.

eandre committed:
introduce duration-based syntax

horiajurcut committed:
Write documentation on how to use Auth Keys (#123)

eandre committed:
docs/auth: add back missing docs (#124)

* docs/auth: add example on returning custom error messages on auth handlers (#79)

* docs: add test and deploy to rest-api tutorial (#100)

* docs/tutorials/rest-api: add missing note on docker

Co-authored-by: wisdommatthew <40186491+wisdommatt@users.noreply.github.com>
Co-authored-by: Torbjörn Einarsson <torbjorn.einarsson@edgeware.tv>

horiajurcut committed:
Add missing docs (#125)

* update missing docs

* fix git command

eandre committed:
docs: update docs on GCP support (#126)

eandre committed:
.github/workflows: update release action to use new encore-go release process (#127)

eandre committed:
docs/configuration/auth-keys: fix encore auth login cli example (#128)

DomBlack committed:
Update release.yml

eandre committed:
.github/workflows: tweak release workflow (#129)

eandre committed:
.: update go.mod to go 1.17 (#130)

DomBlack committed:
Add missing `// +build` tags (#131)

eandre committed:
.github/workflows: bump go version (#132)

DomBlack committed:
Change make release to use softlink for MacOS sdk (#133)

DomBlack committed:
Fix BASE64 Secrets encoding (#135)

This commit fixes the base64 encoding of secrets between the daemon and
run command to both use `RawURLEncoding`. Previously the run command decoded
using `RawStdEncoding` which caused an issue with the secret contained control
characters.

eandre committed:
cli/cmd/encore: improve 'encore db shell' autocompletion (#136)

The 'encore db shell' command takes a service name, but in practice
this is almost always the directory name so it's convenient to be able
to autocomplete the directory names directly.

However most shells autocomplete directories by adding a trailing '/'
which breaks our service name lookup. Handle this better by stripping
any trailing '/'. While we're at it provide a better autocompletion hint
by specifying that only directory names should be autocompleted.

Fixes #103

eandre committed:
docs/develop: document how to handle migration errors

Fixes #68

Thanks to @wisdommatt for starting this effort.

eandre committed:
docs/how-to/share-db-between-services: minor tweaks

Elaborate a bit more on Encore's default stance of having
a database per service, and what the reasons might be for
doing it differently. Finally clean up the code snippet slightly.

eandre committed:
docs/how-to: document how to connect to external dbs

Add a guide on how to use Encore to connect with
external databases and other infrastructure.

horiajurcut committed:
Add support for Every field

eandre committed:
docs/how-to: add missing import to example

eandre committed:
compiler/internal/codegen: support slices in query strings

And add more test cases to catch any future regressions.

horiajurcut committed:
wip

eandre committed:
compiler: fix panic on nested fields in query strings

Return a proper error message instead. To do so, refactor our
use of error lists into a reusable package and use that
in the code generator.

horiajurcut committed:
improve Cron jobs parser

horiajurcut committed:
fix failing parser tests

horiajurcut committed:
validate cron job id

horiajurcut committed:
fix typos

horiajurcut committed:
address pr issues

horiajurcut committed:
fix PR issue
